### ISSUE TITLE
[Question] Functional transforms to a GP posterior

### ISSUE THREAD

**Issue by r-ashwin** [pytorch/botorch#1545]
(apologies if this has been asked before, but I couldn't find anything)
[not a bug. please change label]

Let's say I have a posterior `SingleTaskGP` $f \sim \mathcal{GP}(\mu(x), \sigma^2(x))$. I need to define another `SingleTaskGP` $g \sim \mathcal{GP}(\mu'(x), \sigma'^2(x))$ which is a functional transform of $f$. That is, $\mu'(x) = h(x) \mu(x)$ and $\sigma'^2(x) = h(x)^2 \sigma^2(x)$, where $h(x)$ is a known function.

Can this be done ? If so, I would appreciate a small example to illustrate. Thank you very much!

**Comment by Balandat** [comment]
So we have some functionality for outcome (data-)transforms (https://github.com/pytorch/botorch/blob/main/botorch/models/transforms/outcome.py), which apply a transformation to the observed data points Y (e.g. standardization or log transform) and then apply the reverse transformation to samples or the `Posterior` Object as a whole during prediction.

At least right now though, these operate only on the outputs (and not the location `x`). That could be changed to also take in (optionally) the location `x`, but that would require a good chunk of changes through the code base, believe. But if we had that then this should be straightforward just by implementing the transformation of the posterior. 

If you don't really need to do this generically for multiple model types, one thing you could do would just be to subclass `SingleTaskGP` and overwrite it's `posterior` method to appropriately scale the means and covariance matrix (at this point you do have access to the `x`s .

This would look sth like

```
class Transformed(SingleTaskGP):

def posterior(
        self,
        X: Tensor,
        observation_noise: Union[bool, Tensor] = False,
        posterior_transform: Optional[PosteriorTransform] = None,
        **kwargs: Any,
    ):
    posterior = super().posterior(
        X=X,
        observation_noise=observation_noise,
        posterior_transform=posterior_transform,
        **kwargs,
   )
   scale_fac = h(X)  # here h is your function `h` as a pytorch function that can be evaluated in a vectorized fashion
   new_posterior = scale_posterior(posterior, scale_fac)
   return new_posterior
```

where `scale_posterior` is essentially the same code as https://github.com/pytorch/botorch/blob/main/botorch/models/transforms/outcome.py#L359-L390 (we should probably move that out into some utility to be easier to reuse).

**Comment by r-ashwin** [comment]
Thanks @Balandat .

Just to be clear, I can first train the `SingleTaskGP` $f$, then just instantiating the `TransformedGP` object should give me the posterior I need, correct? The whole point of doing this is to avoid training two GPs instead of one.

**Comment by saitcakmak** [comment]
The `posterior` is not called while training the GP. So it should be fine to just instantiate a `TransformedGP` and train it with `fit_gpytorch_mll` etc. 

What Max suggest is a lot like what we do with the `PosteriorTransform`s, but those currently do not accept `X`. (you can these under botorch/acquisition/objective).

**Comment by saitcakmak** [comment]
> Yeah. Do you foresee any issues with changing the API to also accept an (optional) X in PosteriorTransform.forward()? That would be a more general solution.

I don't see any harm in that. It'd extend the possible use cases and be ignored otherwise.

**Comment by r-ashwin** [comment]
Sounds good.

**Comment by Balandat** [comment]
> We could just store the untrasformed version and pass that to the posterior transform, or we could wait until the refactor is done to avoid duplicate work

Yeah let's wait until the input transform refactor is done. 

@r-ashwin are you unblocked in the mean time with the above suggestions?

**Comment by r-ashwin** [comment]
Yes, thank you very much!

**Comment by Balandat** [comment]
Great, closing this out and moving the feature request to a new issue.
